src/app/services/baw-api/verification/verification.service.ts

```diff
@@ -0,0 +1,201 @@
+      filters
+    );
+  }
+
+  public show(model: IdOr<Verification>): Observable<Verification> {
+    return this.api.show(Verification, endpointShallow(model, emptyParam));
+  }
+
+  public create(model: Verification): Observable<Verification> {
+    return this.api.create(
+      Verification,
+      endpointShallow(emptyParam, emptyParam),
+      (verification) => endpointShallow(verification, emptyParam),
+      model
+    );
+  }
+
+  public update(model: Verification): Observable<Verification> {
+    return this.api.update(
+      Verification,
+      endpointShallow(model, emptyParam),
+      model
+    );
+  }
+
+  public destroy(model: IdOr<Verification>): Observable<void | Verification> {
+    return this.api.destroy(endpointShallow(model, emptyParam));
+  }
+
+  /**
+   * Creates a verification model if it doesn't already exist, if it already
+   * exists, update the existing model.
+   */
+  public createOrUpdate(
+    model: Verification,
+    audioEvent: AudioEvent,
+    user: User
+  ): Observable<Verification> {
+    return this.api
+      .create(
+        Verification,
+        endpointShallow(emptyParam, emptyParam),
+        (verification) => endpointShallow(verification, emptyParam),
+        model,
+        { disableNotification: true }
+      )
+      .pipe(
+        // fetching the verification model here is the only way to be certain
+        // that there are no race conditions
+        catchError((err) => {
+          if (err.status === CONFLICT) {
+            const verificationModel = this.audioEventUserVerification(
+              audioEvent,
+              user
+            );
+            return verificationModel.pipe(
+              mergeMap((verification) => {
+                if (!verification) {
+                  throw err;
+                }
+
+                const newModel = new Verification({
+                  ...verification,
+                  ...model,
+                });
+
+                return this.update(newModel);
+              })
+            );
+          }
+
+          throw err;
+        })
+      );
+  }
+
+  /**
+   * Fetches a users verification model for a specific audio event
+   *
+   * @returns
+   * A verification model if the user has verified the audio event.
+   * If the user has not verified the audio event, null is returned.
+   */
+  public audioEventUserVerification(
+    event: AudioEvent,
+    user: User
+  ): Observable<Verification | null> {
+    const filter = {
+      filter: {
+        and: [
+          { audioEventId: { eq: event.id } },
+          { creatorId: { eq: user.id } },
+        ],
+      },
+    } as const satisfies Filters<Verification>;
+
+    // the api enforces having one verification per user per audio event
+    // therefore, it is safe to assume that there will only be one result
+    // and return the first element of the array
+    return this.filter(filter).pipe(
+      map((results) => (results.length > 0 ? results[0] : null))
+    );
+  }
+}
```

Comment:
i mean you did your best, but oof, having to do a filter here sucks

you can limit the page size to 1 thought - that will help

Reply:
Updated code
![image](https://github.com/user-attachments/assets/f2e173b5-168c-4ebd-a60c-370684110126)